README.md
fah-client-osx-uninstaller
==========================

Folding@home Desktop Client OSX Uninstaller


ChangeLog
v0.1.0
   * Signed, flat package. Requires OSX 10.5+, PPC or Intel.


v0.0.5: (svn 817)
   * Last version for OSX 10.4


notes/scratchpad.sh
#!/bin/bash

exit 1

# notes and experiments for future inclusion in (un)install scripts

#if [ "$UID" -ne 0 ]; then
#    echo "error: you must run this script as root"
#    echo "use"
#    echo "  sudo $0"
#    exit 1
#fi


# fah 5
# future: remove per-user app prefs, for all users
# is this safe? what if some user somehow has a space in their home folder name?
#/bin/rm -f /Users/*/Library/Preferences/edu.stanford.folding.plist >/dev/null 2>&1
#/bin/rm -f /Users/*/Library/Preferences/edu.stanford.folding.viewer.plist >/dev/null 2>&1
# is SUDO_UID set when run by Installer.app?
# we'd like to do this for every user anyway
#sudo -u '#'"$SUDO_UID"  defaults delete edu.stanford.folding
#sudo -u '#'"$SUDO_UID"  defaults delete edu.stanford.folding.viewer
# leave ~/Library/Folding@home alone, so future installs might be able to pick up user,team from client.cfg


# fah 6
# future: maybe kill all fah6, even if run by third-party apps
# or have optional specific detect-and-uninstall for InCrease, finstall, ...
# leave ~/Library/Folding@home alone, so future installs might be able to pick up user,team,passkey from client.cfg


# fah 7 not removed:
# fahclient work directories, data and config files
# global and per-user crash logs, caches
#   /Library/Logs/{CrashReporter,DiagnosticReports}/{FAHClient_,FahCore_}*
#   ~/Library/Logs/{CrashReporter,DiagnosticReports}/{FAHClient_,FahCore_,fahcontrol_}*
#   ~/Library/Caches/<bundle-id>/Cache.db
#   ~/Library/Application Support/CrashReporter/fahcontrol_*
# per-user fahcontrol/fahviewer config files
#   ~/Library/Application Support/FAHClient/FAHControl.db
#   ? where are fahviewer prefs ?


# future:
# stop any fahclient running as a LaunchAgent to fold GPU units
# note: this would not catch any per-user LaunchAgents nor LoginItems
# note: agent label and filename expected to differ from daemon
F="/Library/LaunchAgents/edu.stanford.folding.fahclient.agent.plist"
if [ -f "$F" ]; then
    /bin/launchctl unload -w "$F"
    # note: we should extract fahclient working directory before deleting plists, so we can avoid deleting data
    /bin/rm -f "$F"
fi


# get path to app by bundle id
# this gets one (first?) path, but also launches the app!
osascript -e 'the posix path of (path to application id "edu.stanford.folding.fahcontrol")'


# get all paths to apps with bundle id x
# lsregister works, also works on bundle id fragments due to grep
# NOTE: also can show dups, outdated stuff from cache, and non-app stuff

/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/LaunchServices.framework/Versions/A/Support/lsregister -dump | grep --before-context=2 "edu.stanford.folding.fahcontrol" | grep -v Volumes | grep path: | awk '{print $2}'

/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/LaunchServices.framework/Versions/A/Support/lsregister -dump | grep --before-context=2 "edu.stanford.folding" | grep -v Volumes | grep path: | awk '{print $2}'
# results of second example:
# /Applications/FAHControl.app
# /Users/kevin2/Applications/FAH/Folding@home.app
# /Applications/fahclient.app
# /Users/kevin2/.Trash/Folding@home
# /Users/kevin2/.Trash/Folding@home
# /Applications/fahviewer.app
# /Users/kevin2/.Trash/Folding@home


# for each user, do something
# this doesn't work if there are spaces in folder names
# folder name is also not guaranteed to be user name
USERS=`(cd /Users ; echo * )`
# on newer verions of osx, we could get user names via dscl
# dscl also gets all the hidden _* users, plus root daemon nobody; and uses w/o a home dir
#USERS=`dscl . list /users`
for SOME_USER in $USERS ; do
    if [ "$SOME_USER" != "Guest" ] && [ "$SOME_USER" != "Shared" ] ; then
        echo \"$SOME_USER\"
        # NOTE: must be root for this to work and folder name needs to be same as user name
        if [ -d "/Users/$SOME_USER/Library/Preferences" ] ; then
            sudo -u "$SOME_USER" defaults delete edu.stanford.folding.nonesuch
        fi
    fi
done


exit 0


src/flatdistpackager.py
# flatdistpackager.py
# This should be integrated with packager.py

from SCons.Script import *

deps = ['packager', 'flatdistpkg']

def FlatDistPackager(env, name, **kwargs):
    # call old Packager if not building darwin flat dist
    distpkg_flat = kwargs.get('distpkg_flat', env.get('distpkg_flat', False))
    pkg_type = kwargs.get('pkg_type', env.get('pkg_type'))
    if env['PLATFORM'] != 'darwin' or pkg_type != 'dist' or not distpkg_flat:
        return env.Packager(name, **kwargs)

    # Setup env
    kwargs['package_name'] = name
    kwargs['package_name_lower'] = name.lower()
    env = env.Clone()
    env.Replace(**kwargs)

    # Compute package file name; eg name_version_arch.pkg.zip
    # flat dist packages should be pkg, not mpkg
    # installer doesn't seem to care, currently
    target = env.GetPackageName(name)

    return env.FlatDistPkg(target, [], **kwargs)

def configure(conf):
    env = conf.env
    AddMethod(Environment, FlatDistPackager, 'FlatDistPackager')
    return True


Scripts/uninstall-fah5.sh
#!/bin/bash

# uninstall fah 5

# although fah5 is PPC only, cruft may exist on new machines due to Migration
# Assistant or disk cloning

# stop folding app
killall -TERM "Folding@home" >/dev/null 2>&1

# do not stop nor delete CLI client fah5, as that is assumed to be under manual
# or third-party control

# assume fah screen saver is neither selected nor running
# it would probably have to be a ppc mac, 
# and it is unlikely someone would activate screen saver while uninstaller is
# running

# remove fah5 app
# note that this could have been placed anywhere by user; we'll assume they put
# it in /Apps in the future, we should find and delete all instances on boot
# volume don't delete fake fah6 app
if [ ! -f "/Applications/Folding@home.app/fah6" ]; then
    rm -rf "/Applications/Folding@home.app" >/dev/null 2>&1
fi

# remove screen saver and package receipts

# screen saver used same pkg id (edu.stanford.folding) as later used by fah6
# so only remove 10.6 pkg receipt if saver exists in standard location
# note that it was allowed to install the saver under a user's home; we'll
# ignore that
# in the future, we should find and delete all instances on boot volume
F="/Library/Screen Savers/Folding@home.saver"
if [ -d "$F" ]; then
    # remove screen saver
    rm -rf "$F"
    # remove osx 10.6+ package receipts
    OS_MAJOR="`/usr/bin/uname -r | cut -f1 -d.`"
    if [ "$OS_MAJOR" -ge 10 ]; then
        # We have Snow Leopard or higher.
        pkgutil --force --forget edu.stanford.folding
    fi
fi

# remove pre-10.6 package receipts
rm -rf /Library/Receipts/OSX-SS-5.02.pkg >/dev/null 2>&1

exit 0


Scripts/uninstall-fah6.sh
#!/bin/bash

# uninstall fah 6

# quit System Preferences and remove folding pref pane
if [ -d "/Library/PreferencePanes/Folding@home.prefPane" ]; then
    killall -TERM "System Preferences" >/dev/null 2>&1
    rm -rf "/Library/PreferencePanes/Folding@home.prefPane"
fi

# remove pref pane helper
launchctl unload -w \
    /Library/LaunchDaemons/edu.stanford.folding.prefpane.plist >/dev/null 2>&1
rm -f /Library/LaunchDaemons/edu.stanford.folding.prefpane.plist \
    >/dev/null 2>&1
rm -f /Library/PrivilegedHelperTools/edu.stanford.folding.prefpane \
    >/dev/null 2>&1
rm -f /var/run/edu.stanford.folding.prefpane.socket >/dev/null 2>&1

# stop folding and remove launchd plist
F="/Library/LaunchDaemons/Folding@home.plist"
if [ -f "$F" ]; then
    launchctl unload -w "$F" >/dev/null 2>&1
    rm -f "$F"
fi

# don't kill all fah6, because anything still running probably belongs to a
# third-party tool or was launched manually

# future: kill any true orphan cores and mpiexec, by checking parent process,
# ideally after at least 1 minute
# there can be other mpiexec, so never kill that unless it's parent to a fahcore
# don't kill cores running under fah 7

# remove fah6 client executables
rm -rf /usr/local/fah/ >/dev/null 2>&1

# remove old fake app, which only contained older executables
if [ -x "/Applications/Folding@home.app/fah6" ]; then
    rm -rf "/Applications/Folding@home.app"
fi

# remove osx 10.6+ package receipts
OS_MAJOR="`uname -r | cut -f1 -d.`"
if [ "$OS_MAJOR" -ge 10 ]; then
    # We have Snow Leopard or higher.
    pkgutil --force --forget \
        edu.stanford.folding.foldinghome6241.root.pkg
    pkgutil --force --forget edu.stanford.folding
fi

# remove pre-10.6 package receipts
# old fah6 pref pane installers made root.pkg
F="/Library/Receipts/root.pkg"
if [ -f "$F/Contents/Info.plist" ]; then
    BID=`defaults read "$F/Contents/Info" CFBundleIdentifier`
    # goofy comparison, but it's backwards compatible with old sh
    if [ "$(echo $BID | cut -c1-13)" == "edu.stanford." ] ; then
        rm -rf "$F"
    fi
fi
# these two are theoretical
rm -rf /Library/Receipts/edu.stanford.folding.foldinghome6241.root.pkg \
    >/dev/null 2>&1
rm -rf /Library/Receipts/edu.stanford.folding.pkg >/dev/null 2>&1

exit 0


Scripts/uninstall-cleanup.sh
#!/bin/bash

# uninstaller cleanup

# We must do a delayed cleanup, so we can remove the uninstaller receipts,
# which are written AFTER all postflights.
# Metapkg postflight is actually run before the component postflights.
if [ "$1" != "--delayed-cleanup" ]; then
  "$0" --delayed-cleanup "$@" &
  exit 0
fi
shift

# Wait up to 1 minute for Installer.app to quit
I=0
while $(ps axww|grep -v grep|grep Installer.app >/dev/null 2>&1); do
  sleep 1
  let I+=1
  if [ $I -ge 60 ]; then break; fi
done

if [ $I -lt 1 ]; then sleep 1; fi

# remove uninstaller package receipts
# edu.stanford.folding.uninstaller.pkg
# edu.stanford.folding.uninstall.fah{5,6,7}.pkg
# edu.stanford.folding.uninstall.cleanup.pkg

# derived from vmware uninstaller script
# remove osx 10.6+ package receipts
OS_MAJOR="`/usr/bin/uname -r | cut -f1 -d.`"
if [ "$OS_MAJOR" -ge 10 ]; then
   # We have Snow Leopard or higher.
   PACKAGES=`pkgutil --pkgs='edu\.stanford\.folding\.uninstall.*'`
   OLD_IFS="$IFS"
   IFS=$'\n'
   # Discard the existing package receipt data including any old pkg ID(s).
   for pkg in $PACKAGES ; do
      pkgutil --force --forget $pkg
   done
   IFS="$OLD_IFS"
   # We don't use pkgutil --unlink to remove the installed package contents
   # because it does not remove the directories. We fall back to removing
   # the files manually below. If pkgutil is ever updated to remove the
   # directories, we should use it instead, and remove any install/runtime
   # created files here.
fi

# remove pre-10.6 package receipts
rm -rf /Library/Receipts/edu.stanford.folding.uninstall* >/dev/null 2>&1

# finally, remove uninstaller pkg, from which we may have been run
# Installer.app may move or delete this early on its own
rm -rf "/Applications/Folding@home/Uninstall Folding@home.pkg" >/dev/null 2>&1

# remove folder if empty
rm -f /Applications/Folding\@home/.DS_Store >/dev/null 2>&1
rmdir /Applications/Folding\@home >/dev/null 2>&1

exit 0


Scripts/uninstall-fah7.sh
#!/bin/bash

# uninstall fah 7

# terminate FAHControl and FAHViewer if running; also terminate by old exe names
# this should also stop any fahclient running under fahcontrol
# NO, will not stop client, as SIGTERM is not caught for a normal exit
killall -TERM FAHControl FAHViewer fahcontrol FAHClientGUI >/dev/null 2>&1

# stop folding service; old plist name
F="/Library/LaunchDaemons/FAHClient.plist"
if [ -f "$F" ]; then
    launchctl unload -w "$F"
    # note: we should extract fahclient working directory before deleting
    # plists, so we can avoid deleting data
    rm -f "$F"
fi

# stop folding service
F="/Library/LaunchDaemons/edu.stanford.folding.fahclient.plist"
if [ -f "$F" ]; then
    launchctl unload -w "$F"
    rm -f "$F"
else
    # try to unload job, even if user manually deleted plist
    launchctl remove edu.stanford.folding.fahclient >/dev/null 2>&1
fi

# remove client helper agent plist
F="/Library/LaunchAgents/edu.stanford.folding.fahclienthelper.plist"
if [ -f "$F" ]; then
    # don't bother to unload from all users
    # harmless to leave running till reboot
    rm -f "$F"
fi

# future: stop any fahclient running as root or per-user LaunchAgent, or
# per-user LoginItem to fold GPU units, or per-user via SpawnApps

# kill any FAHClient still running
if ps auxwww | grep "FAHClient" | grep -v 'grep' >/dev/null 2>&1; then
    # wait a second for client to stop on its own, then TERM any still running
    sleep 1
    killall -TERM FAHClient >/dev/null 2>&1
fi

# trust that all cores and corewrappers will quit normally

# remove scripts, symlinks, executables
rm -f /usr/bin/FAHControl >/dev/null 2>&1
rm -f /usr/bin/FAHViewer >/dev/null 2>&1
rm -f /usr/bin/FAHClient >/dev/null 2>&1
rm -f /usr/bin/FAHCoreWrapper >/dev/null 2>&1
rm -f /usr/bin/FAHClientHelper >/dev/null 2>&1

# remove any fahclient extras
# note, in the future, this might include fahclient working directory
# so we really should check for and not delete any such files
rm -rf /usr/lib/fahclient/ >/dev/null 2>&1

# remove apps, by old and new names
# future: also find and delete any moved copies of the apps
rm -rf /Applications/FAHClientGUI.app >/dev/null 2>&1
rm -rf /Applications/Folding\@home\ Client.app >/dev/null 2>&1

rm -rf /Applications/fahclient.app >/dev/null 2>&1
rm -rf /Applications/fahviewer.app >/dev/null 2>&1

rm -rf /Applications/FAHControl.app >/dev/null 2>&1
rm -rf /Applications/FAHViewer.app >/dev/null 2>&1
rm -f /Applications/FAHClient.url >/dev/null 2>&1

rm -rf /Applications/Folding\@home/FAHControl.app >/dev/null 2>&1
rm -rf /Applications/Folding\@home/FAHViewer.app >/dev/null 2>&1
rm -f /Applications/Folding\@home/FAHClient.url >/dev/null 2>&1
rm -f /Applications/Folding\@home/Web\ Control.url >/dev/null 2>&1

# remove fahcontrol helper, if it exists
F="/Library/LaunchDaemons/edu.stanford.folding.fahcontrol.helper.plist"
if [ -f "$F" ]; then
    launchctl unload -w "$F"
    rm -f "$F"
    rm -f \
        /Library/PrivilegedHelperTools/edu.stanford.folding.fahcontrol.helper
    rm -f /var/run/edu.stanford.folding.fahcontrol.helper.socket
fi
# unpatched BetterhAuthenticationSampleLib doesn't append .helper to bundleid
F="/Library/LaunchDaemons/edu.stanford.folding.fahcontrol.plist"
if [ -f "$F" ]; then
    launchctl unload -w "$F"
    rm -f "$F"
    rm -f /Library/PrivilegedHelperTools/edu.stanford.folding.fahcontrol
    rm -f /var/run/edu.stanford.folding.fahcontrol.socket
fi

# remove osx 10.6+ package receipts
OS_MAJOR="`uname -r | cut -f1 -d.`"
if [ "$OS_MAJOR" -ge 10 ]; then
    # We have Snow Leopard or higher.
    pkgutil --force --forget edu.stanford.folding.fahclient.pkg
    pkgutil --force --forget edu.stanford.folding.fahcontrol.pkg
    pkgutil --force --forget edu.stanford.folding.fahviewer.pkg
fi

# remove pre-10.6 package receipts
rm -rf /Library/Receipts/edu.stanford.folding.fahclient.pkg \
    >/dev/null 2>&1
rm -rf /Library/Receipts/edu.stanford.folding.fahcontrol.pkg \
    >/dev/null 2>&1
rm -rf /Library/Receipts/edu.stanford.folding.fahviewer.pkg \
    >/dev/null 2>&1
# on 10.4, I get these
rm -rf /Library/Receipts/fahclient.pkg >/dev/null 2>&1
rm -rf /Library/Receipts/fahcontrol.pkg >/dev/null 2>&1
rm -rf /Library/Receipts/fahviewer.pkg >/dev/null 2>&1

exit 0
